internal/ingress/controller/controller.go

```diff
@@ -226,25 +230,47 @@ type KongController struct {
 	PluginSchemaStore PluginSchemaStore
 }
 
-// Start start a new NGINX master process running in foreground.
+// Start starts a new NGINX master process running in foreground, blocking until the next call to
+// Stop.
 func (n *KongController) Start() {
 	glog.Infof("starting Ingress controller")
 
-	go n.elector.Run()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		<-n.stopCh
+		cancel()
+	}()
+
+	goTallied := func(f func()) {
+		n.wgBackgroundActive.Add(1)
+		go func() {
+			defer n.wgBackgroundActive.Done()
+			f()
+		}()
+	}
+
+	goTallied(func() {
+		n.elector.Run(ctx)
+	})
 
 	if n.syncStatus != nil {
-		go n.syncStatus.Run()
+		goTallied(func() {
+			n.syncStatus.Run()
+		})
 	}
 
-	go n.syncQueue.Run(time.Second, n.stopCh)
-	// force initial sync
+	goTallied(func() {
+		n.syncQueue.Run(time.Second, n.stopCh)
+	})
```

Comment:
Can you simplify this code by using https://godoc.org/golang.org/x/sync/errgroup package?
There is no need of collecting errors but other than that, errgroup package can make this simpler.

Reply:
Yes, [`(*errgroup.Group).Go` is `goTallied`](https://github.com/golang/sync/blob/master/errgroup/errgroup.go#L51-L66), with the extra error collection. In our case, none of these functions return errors, so we still need to wrap them.

Please see commit bc353043b503ff2062ce6d1786eab508d5b46d80 and let me know if that's more to your liking.